lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskRuntime.cs

```diff
@@ -92,50 +93,67 @@ public System.Threading.Tasks.Task RunTask()
 
             // Send heartbeat such that user receives a TaskRunning message.
             _currentStatus.SetRunning();
-            
+
             return System.Threading.Tasks.Task.Run(() =>
             {
                 Logger.Log(Level.Info, "Calling into user's task.");
                 return _userTask.Call(null);
-            }).ContinueWith(runTask =>
+            }).ContinueWith((System.Threading.Tasks.Task<byte[]> runTask) =>
+            {
+                try
                 {
-                    try
+                    // Task failed.
+                    if (runTask.IsFaulted)
                     {
-                        // Task failed.
-                        if (runTask.IsFaulted)
-                        {
-                            Logger.Log(Level.Warning,
-                                string.Format(CultureInfo.InvariantCulture, "Task failed caused by exception [{0}]", runTask.Exception));
-                            _currentStatus.SetException(runTask.Exception);
-                            return;
-                        }
+                        OnTaskFailure(runTask);
+                        return;
+                    }
 
-                        if (runTask.IsCanceled)
-                        {
-                            Logger.Log(Level.Warning,
-                                string.Format(CultureInfo.InvariantCulture, "Task failed caused by task cancellation"));
-                            return;
-                        }
+                    if (runTask.IsCanceled)
+                    {
+                        Logger.Log(Level.Error,
+                            string.Format(CultureInfo.InvariantCulture, "Task failed caused by System.Threading.Task cancellation"));
+                        OnTaskFailure(runTask);
+                        return;
+                    }
 
-                        // Task completed.
-                        var result = runTask.Result;
-                        Logger.Log(Level.Info, "Task Call Finished");
-                        _currentStatus.SetResult(result);
-                        if (result != null && result.Length > 0)
-                        {
-                            Logger.Log(Level.Info, "Task running result:\r\n" + System.Text.Encoding.Default.GetString(result));
-                        }
+                    // Task completed.
+                    var result = runTask.Result;
+                    Logger.Log(Level.Info, "Task Call Finished");
+                    _currentStatus.SetResult(result);
+                    if (result != null && result.Length > 0)
+                    {
+                        Logger.Log(Level.Info, "Task running result:\r\n" + System.Text.Encoding.Default.GetString(result));
```

Comment:
While you are at it, please move this to a higher log level and make the string construction conditional on the log level. For applications with many short-lived tasks, this will be a performance issue otherwise.